File: codigo/matlab/diferencias_finitas.m
function [d1, d2] = diferencias_finitas(f, x, h)
% DIFERENCIAS_FINITAS  Derivadas por diferencias finitas centradas. Chapra cap. 23.
%   d1 : primera derivada,  d2 : segunda derivada
%   Uso:  [d1, d2] = diferencias_finitas(@sin, 1.0);
    if nargin < 3, h = 1e-5; end
    d1 = (f(x+h) - f(x-h))/(2*h);
    d2 = (f(x+h) - 2*f(x) + f(x-h))/h^2;
end

File: codigo/python/diferencias_finitas.py
"""diferencias_finitas.py -- Derivacion numerica por diferencias finitas. Chapra cap. 23."""
import math


def adelante(f, x, h=1e-5):
    return (f(x + h) - f(x)) / h


def atras(f, x, h=1e-5):
    return (f(x) - f(x - h)) / h


def centrada(f, x, h=1e-5):
    return (f(x + h) - f(x - h)) / (2 * h)


def segunda(f, x, h=1e-4):
    return (f(x + h) - 2 * f(x) + f(x - h)) / h**2


if __name__ == "__main__":
    f, x = math.sin, 1.0
    print(f"f'(1)  centrada = {centrada(f, x):.8f}   (cos 1 = {math.cos(1):.8f})")
    print(f"f''(1)          = {segunda(f, x):.8f}   (-sin 1 = {-math.sin(1):.8f})")
